src/everwillow/statelib/meta.py

```diff
@@ -0,0 +1,103 @@
+from __future__ import annotations
+
+import dataclasses
+import typing as tp
+from functools import partial
+
+import jax.tree_util as jtu
+from jaxtyping import PyTree, PyTreeDef
+
+if tp.TYPE_CHECKING:
+    from everwillow.statelib.state import FrozenChainMap, K, State, V
+
+
+__all__ = ["MergeMeta", "TreeDefMeta"]
+
+
+@partial(
+    jtu.register_dataclass,
+    data_fields=[],
+    meta_fields=["treedef", "keys"],
+)
+@dataclasses.dataclass(frozen=True, slots=True)
+class TreeDefMeta:
+    """Metadata retaining the treedef of a pytree."""
+
+    treedef: PyTreeDef | None
+    keys: tp.Sequence[K]
+
+    def to_pytree(
+        self,
+        mapping: tp.Mapping[K, V],
+    ) -> PyTree[V]:
+        """Reconstruct a ``State`` from flattened values.
+
+        Args:
+            values: Flattened values corresponding to the original pytree.
+
+        Returns:
+            Reconstructed ``State`` object.
+        """
+        # can't reconstruct pytree with treedef=None
+        if self.treedef is None:
+            msg = "Cannot reconstruct pytree with 'treedef=None'"
+            raise ValueError(msg)
+
+        # can't convert to pytree if keys don't match
+        self_keys = set(self.keys)
+        mapping_keys = set(mapping.keys())
+        if self_keys != mapping_keys:
+            missing = self_keys - mapping_keys
+            extra = mapping_keys - self_keys
+            msg = f"Mapping keys do not match treedef keys. Missing: {missing}, Extra: {extra}"
+            raise KeyError(msg)
+
+        del self_keys, mapping_keys  # unused after check
+
+        # this order of keys is important here to preserve original pytree order
+        return jtu.tree_unflatten(
+            treedef=self.treedef, leaves=(mapping[k] for k in self.keys)
+        )
+
+
+@partial(
+    jtu.register_dataclass,
+    data_fields=[],
+    meta_fields=["treedefmetas"],
+)
+@dataclasses.dataclass(frozen=True, slots=True)
+class MergeMeta:
+    """Metadata retained when multiple :class:`State` objects are merged.
+
+    The metadata stores the treedef for each original state alongside the group
+    of keys that belong to that state so the merged mapping can be split later.
```

Comment:
it only stores the treedef, not keys, right? the treedefs contain the key information

Reply:
the 'treedefmeta' tracks treedefs (from pytrees) and an ordered iterable of keys